core/src/services/supabase/core.rs

```diff
@@ -0,0 +1,290 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::Debug;
+
+use anyhow::anyhow;
+use bytes::Bytes;
+use http::header::CONTENT_LENGTH;
+use http::header::CONTENT_TYPE;
+use http::HeaderValue;
+use http::Request;
+use http::Response;
+use serde_json::json;
+
+use crate::raw::*;
+use crate::*;
+
+mod constants {
+    pub const AUTH_HTTP_HEADER_KEY: &str = "authorization";
+    pub const SUPABASE_AUTH_ANON_KEY: &str = "OPENDAL_SUPABASE_AUTH_ANON_KEY";
+    pub const SUPABASE_AUTH_SERVICE_KEY: &str = "OPENDAL_SUPABASE_AUTH_SERVICE_KEY";
+}
+
+pub struct SupabaseCore {
+    pub root: String,
+    pub bucket: String,
+    pub endpoint: String,
+
+    /// The key used for authorization, initialized by environment variable you designated.
+    /// Normally it is rather an anon_key(Client key) or an service_role_key(Secret Key)
+    pub auth_key: Option<HeaderValue>,
+    /// This is true if the service_role_key is loaded, false if the anon_key is loaded
+    pub auth: bool,
+
+    pub http_client: HttpClient,
+}
+
+impl Debug for SupabaseCore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SupabaseCore")
+            .field("root", &self.root)
+            .field("bucket", &self.bucket)
+            .field("endpoint", &self.endpoint)
+            .finish_non_exhaustive()
+    }
+}
+
+impl SupabaseCore {
+    pub fn new(root: &str, bucket: &str, endpoint: &str, client: HttpClient) -> Self {
+        Self {
+            root: root.to_string(),
+            bucket: bucket.to_string(),
+            endpoint: endpoint.to_string(),
+            auth_key: None,
+            auth: false,
+            http_client: client,
+        }
+    }
+
+    /// load auth key tries to load the authorization key from the environment variable
+    /// - if the key is set by user using Builder, this return directly
+    /// - if the key is not set by user, this will try to load the service key first, then the anon key
+    pub fn load_auth_key(&mut self) {
```

Comment:
We don't need this

Reply:
> If so, how and where to get those envs?

We don't need to got those envs. Users should not configure opendal via `OPENDAL_XXXX`. We only load services' it own env like `AWS_ACCESS_KEY_ID`.

Reply:
> > We don't need this
> 
> If so, how and where to get those envs?

Probably w/ the code like
```Rust
let key = std::env::var("SOME_USER_ASSIGNED_ENV");
builder.anon_key(key);
```

Reply:
A service should not declare it will need env like `OPENDAL_XXX`. It doesn't work like this.

---

Service should declare it need config `abc`, and

- provide a public function called `abc()`.
- load `abc` in `from_map` and call `abc()` to set it.
- And finlly, `from_env` will load all `OPENDAL_SUPABASE_XXX` from env and call `from_map`.

Reply:
> ```rust
> let key = std::env::var("SOME_USER_ASSIGNED_ENV");
> builder.anon_key(key);
> ```

It's our application developers own decision and not related to opendal.

Reply:
> > ```rust
> > let key = std::env::var("SOME_USER_ASSIGNED_ENV");
> > builder.anon_key(key);
> > ```
> 
> It's our application developers own decision and not related to opendal.

Yea, that's what I mean, it is the user's own choice on how to call on our key() interface

Reply:
What I want to say is that, if these backends do not use `reqsign`, then they lose the ability to load properties from the environment variable, right? Perhaps our past experience with using services like S3 may have caused some confusion.

So users need to handle this process themselves. We may need to write this behavior into our docs.

Reply:
> What I want to say is that, if these backends do not use `reqsign`, then they lose the ability to load properties from the environment variable, right?

Supabase doesn't have properties to load from env. It's a simple service which only requires a url and key.

Reply:
> What I want to say is that, if these backends do not use `reqsign`, then they lose the ability to load properties from the environment variable, right? Perhaps our past experience with using services like S3 may have caused some confusion.
> 
> So users need to handle this process themselves. We may need to write this behavior into our docs.

I think we could write authorization key's behaviors in detail in the rust-doc.

Reply:
> > I think we could write authorization key's behaviors in detail in the rust-doc.
> 
> We don't need to write docs for not existing things...

I mean documenting on the differences of anon_key and service_key.

Reply:
> I mean documenting on the differences of anon_key and service_key.

Oh, sorry. I thought we still talking about the env loading logic.

And yes, we need to let users know about the difference of `anon_key` and `service_key`.

Reply:
> OpenDAL will load config from env if `service providers` declares their standards like `AWS_ACCESS_KEY_ID`. And OpenDAL will not expose it's internal env to users.

Got it.

Reply:
> Instead, we can give developers some notice when implementing backend.

Nice idea. Maybe we can add some notes in the super duck service docs.